extrafiles/tooling/nktooling/__init__.py

```diff
@@ -0,0 +1,129 @@
+	"""
+	return READONLY
+
+def failIfReadonly(prgname):
+	"""
+	Helper function indented to give a summary after a test in read-only mode.
+	In read-only mode, fixes could not be applied, hence, we fail with a negative
+	return code. This can be used during CI to check whether repository is compliant.
+	"""
+	if not isReadonly():
+		return
+
+	print("Changes are necessary.")
+	print("Use '%s -w' to apply changes." % prgname)
+	sys.exit(-1)
+
+def getNetworKitRoot():
+	"""
+	Returns an absolute path to the project root (which is derived from the module's absolute path)
+	"""
+	return os.path.realpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", "..", ".."))
+
+def getCXXFiles(includeSources = True, includeHeaders = True):
+	"""
+	Returns a list of all NetworKit C++ source files.
+
+	includeSource : bool
+		controls whether files from the source directory (networkit/cpp) are included.
+	includeHeaders : bool
+		controls whether file from the include directory (include/networkit) are included.
+	"""
+	files = []
+	if includeSources:
+		newfiles = glob.glob(os.path.join(getNetworKitRoot(), "networkit", "cpp", "**", "*.[ch]pp"), recursive=True)
+		assert(len(newfiles) > 0)
+		files += newfiles
+
+	if includeHeaders:
+		newfiles = glob.glob(os.path.join(getNetworKitRoot(), "include", "networkit", "**", "*.[ch]pp"), recursive=True)
+		assert(len(newfiles) > 0)
+		files += newfiles
+
+	return files
+
+class FileRewriter:
+	"""
+	The FileRewriter acts as an wrapper to safely
+	read and update a text file provided to the constructor.
+	To read the file iterate for lines().
+	To write a (un)modified line call write(line).
+	Once your done, call commit() to replace the file atomically.
+	Observe that commit will do nothing if the nktooling module
+	is in readonly mode (which is the default).
+	"""
+	def __init__(self, filepath):
+		self.path = filepath
+		self.in_lines = []
+		self.out_lines = []
+		self.reading = False
+
+	def lines(self):
+		"""
+		Iterates over all input lines.
+		"""
+		with open(self.path, "r") as fh:
+			self.reading = True
+			for line in fh:
+				self.in_lines.append(line)
+				yield line
+
+			self.reading = False
+
+	def write(self, line):
+		"""
+		Adds data to the output buffer.
+		"""
+		self.out_lines.append(line)
+
+	def isIdentical(self):
+		"""
+		Checks whether the content written matches the content read (so far).
+		"""
+		return self.in_lines == self.out_lines
+
+	def commit(self, force = False):
+		"""
+		Writes buffered lines into the specified in the constructor.
+		Writes are carried out atomically, i.e. we first write into a temporary file
+		and then replace the original.
+		If the module is currently readonly (which is the default, see setup) no change
+		is carried out.
+		This function may only be called if not reads are taking place currently (e.g.,
+		via lines())
+		"""
+		assert(not self.reading)
+		if isReadonly() and not force:
+			return
+
+		with tempfile.TemporaryDirectory() as dir:
```

Comment:
We create an entirely new directory for each file here. Why don't we just create a temporary file?

Reply:
For me this looked like the cleanest approach (I do really think that the cost of creating a directory is insignificant compared to a Python parsing source code ;) to make sure
0.) we get a unique path
1.) the file is automatically cleaned up
2.) we can move the file out of it.

I'm aware only of one other variant: `NamedTemporaryFile`; however here we do [not get the guarantee](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile) that the file can be (i.e., moved) while still being opened. Do you know a better alternative?

Reply:
Hm, I was not aware of this limitation. In this case the temporary directory indeed looks like a good solution.